plugin/pkg/shoot/resourcereservation/admission.go

```diff
@@ -0,0 +1,295 @@
+
+	for i := 0; i < len(shoot.Spec.Provider.Workers); i++ {
+		workerPath := workersPath.Index(i)
+		worker := &shoot.Spec.Provider.Workers[i]
+
+		allErrs = injectResourceReservations(worker, cloudProfile, *workerPath, allErrs)
+	}
+
+	if len(allErrs) > 0 {
+		return apierrors.NewInvalid(a.GetKind().GroupKind(), a.GetName(), allErrs)
+	}
+
+	return nil
+}
+
+func setStaticResourceReservationDefaults(shoot *core.Shoot) {
+	var (
+		kubeReservedMemory = resource.MustParse("1Gi")
+		kubeReservedCPU    = resource.MustParse("80m")
+		kubeReservedPID    = resource.MustParse("20k")
+	)
+
+	if shoot.Spec.Kubernetes.Kubelet == nil {
+		shoot.Spec.Kubernetes.Kubelet = &core.KubeletConfig{}
+	}
+	kubelet := shoot.Spec.Kubernetes.Kubelet
+
+	if kubelet.KubeReserved == nil {
+		kubelet.KubeReserved = &core.KubeletConfigReserved{
+			CPU:    &kubeReservedCPU,
+			Memory: &kubeReservedMemory,
+			PID:    &kubeReservedPID,
+		}
+	} else {
+		if kubelet.KubeReserved.Memory == nil {
+			kubelet.KubeReserved.Memory = &kubeReservedMemory
+		}
+		if kubelet.KubeReserved.CPU == nil {
+			kubelet.KubeReserved.CPU = &kubeReservedCPU
+		}
+		if kubelet.KubeReserved.PID == nil {
+			kubelet.KubeReserved.PID = &kubeReservedPID
+		}
+	}
+}
+
+func injectResourceReservations(worker *core.Worker, cloudProfile *gardencorev1beta1.CloudProfile, path field.Path, allErrs field.ErrorList) field.ErrorList {
+	reservation, err := calculateResourceReservationForMachineType(cloudProfile, worker.Machine.Type)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(path.Child("machine", "type"), worker.Machine.Type, "worker machine type unknown"))
+		return allErrs
+	}
+
+	if worker.Kubernetes == nil {
+		worker.Kubernetes = &core.WorkerKubernetes{}
+	}
+	if worker.Kubernetes.Kubelet == nil {
+		worker.Kubernetes.Kubelet = &core.KubeletConfig{}
+	}
+	if worker.Kubernetes.Kubelet.KubeReserved == nil {
+		worker.Kubernetes.Kubelet.KubeReserved = reservation
+	} else {
+		kubeReserved := worker.Kubernetes.Kubelet.KubeReserved
+		if kubeReserved.CPU == nil {
+			kubeReserved.CPU = reservation.CPU
+		}
+		if kubeReserved.Memory == nil {
+			kubeReserved.Memory = reservation.Memory
+		}
+		if kubeReserved.PID == nil {
+			kubeReserved.PID = reservation.PID
+		}
+	}
+	return allErrs
+}
+
+func calculateResourceReservationForMachineType(cloudProfile *gardencorev1beta1.CloudProfile,
+	machineType string) (*core.KubeletConfigReserved, error) {
+
+	kubeReservedPID := resource.MustParse("20k")
+	for _, machine := range cloudProfile.Spec.MachineTypes {
+		if machine.Name == machineType {
+			cpuReserved := calculateCPUReservation(machine.CPU.MilliValue())
+			memoryReserved := calculateMemoryReservation(machine.Memory.Value())
+
+			return &core.KubeletConfigReserved{
+				CPU:    resource.NewMilliQuantity(cpuReserved, resource.BinarySI),
+				Memory: resource.NewQuantity(memoryReserved, resource.BinarySI),
+				PID:    &kubeReservedPID,
+			}, nil
+		}
+	}
```

Comment:
Instead of iterating through the potentially long list of machine types again and again for each worker pool and shoot, what about building a `map[string]core.MachineType` upfront so that we can access the machine type data based on the name in `O(1)`?

Reply:
I've added a function `buildMachineTypeMap(cloudProfile *gardencorev1beta1.CloudProfile) map[string]gardencorev1beta1.MachineType`. Currently, that function runs every time the admission hook is called. Or did you have in mind to cache the map? (I'm not sure whether there's already a pattern somewhere in Gardener that would allow doing so in a race-free way?)

Reply:
The function is at https://github.com/gardener/gardener/pull/9449/files#diff-e5db5c7cca19cd1aa25f19d3df2c16e7e96a831ffe5ad82f4f87a40eca5cfa8aR233 .

Reply:
Hm, seems GitHub was showing me some old version of this PR last week. Anyways, thanks!

Reply:
I had the same problem this morning in this PR and ended up totally confused :confused:  .